File: Tarea 2/static/js/validations.js
// validaciones de cliente alineadas con el backend

(function () {
  const form = document.getElementById("form-aviso");
  if (!form) return;

  // regex basicos
  const EMAIL_RE = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
  const CEL_RE = /^\+569\d{8}$/; // formato CL estricto

  // helpers
  const isBlank = (v) => v == null || String(v).trim() === "";
  const intOrNaN = (v) => {
    const n = parseInt(String(v), 10);
    return Number.isNaN(n) ? NaN : n;
  };

  // pinta borde rojo o limpia
  function mark(el, bad) {
    if (!el) return;
    el.style.borderColor = bad ? "red" : "";
  }

  // contenedor de errores alto nivel
  function renderTopErrors(msgs) {
    let box = document.querySelector(".errors");
    if (!box) {
      box = document.createElement("div");
      box.className = "errors";
      form.parentNode.insertBefore(box, form);
    }
    box.innerHTML = "";
    const ul = document.createElement("ul");
    for (const m of msgs) {
      const li = document.createElement("li");
      li.textContent = m;
      ul.appendChild(li);
    }
    box.appendChild(ul);
  }

  // valida el formulario y retorna lista de mensajes
  function validate() {
    const errs = [];

    const region = document.getElementById("region");
    const comuna = document.getElementById("comuna_id");
    const sector = document.getElementById("sector");
    const nombre = document.getElementById("nombre");
    const email = document.getElementById("email");
    const celular = document.getElementById("celular");
    const tipo = document.getElementById("tipo");
    const cantidad = document.getElementById("cantidad");
    const edad = document.getElementById("edad");
    const unidad = document.getElementById("unidad_medida");
    const fecha = document.getElementById("fecha_entrega");
    const desc = document.getElementById("descripcion");

    // reset marcas
    [region, comuna, sector, nombre, email, celular, tipo, cantidad, edad, unidad, fecha, desc].forEach((el) =>
      mark(el, false)
    );

    // requeridos
    const req = [
      ["comuna_id", comuna],
      ["nombre", nombre],
      ["email", email],
      ["tipo", tipo],
      ["cantidad", cantidad],
      ["edad", edad],
      ["unidad_medida", unidad],
      ["fecha_entrega", fecha],
    ];
    const faltan = [];
    for (const [key, el] of req) {
      if (!el || isBlank(el.value)) {
        faltan.push(key);
        mark(el, true);
      }
    }
    if (faltan.length) errs.push("Faltan campos obligatorios: " + faltan.join(", "));

    // longitudes
    if (sector && sector.value && sector.value.length > 100) {
      errs.push("Sector excede 100");
      mark(sector, true);
    }
    if (nombre && nombre.value && nombre.value.length > 200) {
      errs.push("Nombre excede 200");
      mark(nombre, true);
    }
    if (email && email.value && email.value.length > 100) {
      errs.push("Email excede 100");
      mark(email, true);
    }
    if (celular && celular.value && celular.value.length > 12) {
      errs.push("Celular excede 12");
      mark(celular, true);
    }
    if (desc && desc.value && desc.value.length > 500) {
      errs.push("Descripcion excede 500");
      mark(desc, true);
    }

    // formatos
    if (email && !isBlank(email.value) && !EMAIL_RE.test(email.value)) {
      errs.push("Email invalido");
      mark(email, true);
    }
    if (celular && !isBlank(celular.value) && !CEL_RE.test(celular.value)) {
      errs.push("Celular invalido");
      mark(celular, true);
    }
    if (tipo && tipo.value && !["gato", "perro"].includes(tipo.value)) {
      errs.push("Tipo invalido");
      mark(tipo, true);
    }
    if (unidad && unidad.value && !["a", "m"].includes(unidad.value)) {
      errs.push("Unidad invalida");
      mark(unidad, true);
    }

    // enteros y rangos
    const cant = intOrNaN(cantidad && cantidad.value);
    if (Number.isNaN(cant)) {
      errs.push("Cantidad debe ser entero");
      mark(cantidad, true);
    } else if (cant < 1) {
      errs.push("Cantidad debe ser mayor o igual a 1");
      mark(cantidad, true);
    }

    const ed = intOrNaN(edad && edad.value);
    if (Number.isNaN(ed)) {
      errs.push("Edad debe ser entero");
      mark(edad, true);
    } else if (ed < 0) {
      errs.push("Edad no puede ser negativa");
      mark(edad, true);
    }

    // fecha futura estricta
    if (fecha && !isBlank(fecha.value)) {
      const dt = new Date(fecha.value);
      const now = new Date();
      if (!(dt instanceof Date) || Number.isNaN(dt.getTime())) {
        errs.push("Fecha de entrega invalida");
        mark(fecha, true);
      } else if (dt.getTime() <= now.getTime()) {
        errs.push("Fecha de entrega debe ser futura");
        mark(fecha, true);
      }
    }

    return errs;
  }

  // intercepta submit y valida
  form.addEventListener("submit", function (ev) {
    const errs = validate();
    if (errs.length) {
      ev.preventDefault();
      renderTopErrors(errs);
      // desplaza arriba para ver el bloque de errores
      window.scrollTo({ top: 0, behavior: "smooth" });
    }
  });
})();

File: Tarea 3/utils/validations.py
from datetime import datetime
import re
from sqlalchemy import select
from models import Comuna

# expresiones regulares para validar email y celular
EMAIL_RE = re.compile(r"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$")
CELULAR_RE = re.compile(r"^\+569\d{8}$") # formato +569XXXXXXXX


def parse_fecha(fecha_str):
    try:
        return datetime.fromisoformat(fecha_str)
    except Exception:
        return datetime.strptime(fecha_str, "%Y-%m-%d")
    
def validar_email(email):
    # entrega true si el email es valido, false si no
    return EMAIL_RE.match(email) is not None

def validar_celular(celular):
    # entrega true si el celular es valido, false si no
    return CELULAR_RE.match(celular) is not None

def _is_blank(v):
    if v is None:
        return True
    if isinstance(v, str) and v.strip() == "":
        return True
    return False

def validate_form(db, form):
    # creamos una lista vacia para almacenar los errores
    errores = []

    # campos obligatorios
    req = ["comuna_id", "nombre", "email", "tipo", "cantidad", "edad", "unidad_medida", "fecha_entrega"]
    faltan = [k for k in req if _is_blank(form.get(k))]
    if faltan:
        errores.append(f"Faltan campos obligatorios: {', '.join(faltan)}")

    # longitudes segun esquema
    if form.get("sector") and len(form["sector"]) > 100:
        errores.append("El sector no puede tener más de 100 caracteres")
    if form.get("nombre") and len(form["nombre"]) > 200:
        errores.append("El nombre no puede tener más de 200 caracteres")

    # validamos email y celular agregando las funciones auxiliares
    if form.get("email") and (len(form["email"]) > 100 or not validar_email(form["email"])):
        errores.append("El email no es válido o tiene más de 100 caracteres")
    if form.get("celular") and (len(form["celular"]) > 12 or not validar_celular(form["celular"])):
        errores.append("El celular no es válido o tiene más de 12 caracteres")

    if form.get("descripcion") and len(form["descripcion"]) > 500:
        errores.append("La descripción no puede tener más de 500 caracteres")
    if form.get("tipo") and form["tipo"] not in ("gato", "perro"):
        errores.append("El tipo debe ser 'gato' o 'perro'")
    if form.get("unidad_medida") and form["unidad_medida"] not in {"a", "m"}:
        errores.append("La unidad de medida debe ser 'a' (años) o 'm' (meses)")


    # validamos enteros y rangos
    try:
        cantidad = int(form.get("cantidad", ""))
        if cantidad < 1:
            errores.append("La cantidad debe ser mayor o igual a 1")
        
    except Exception:
        errores.append("La cantidad debe ser un entero")

    try: 
        edad = int(form.get("edad", ""))
        if edad < 0:
            errores.append("La edad no puede ser negativa")
    except Exception:
        errores.append("La edad debe ser un entero")

    # validamos que la comuna exista en la base de datos
    try:
        comuna_id = int(form["comuna_id"])
        ok = db.execute(select(Comuna).where(Comuna.id == comuna_id)).scalar_one_or_none()
        if not ok:
            errores.append("La comuna no existe")
    except Exception:
        errores.append("La comuna no es válida")

    # validamos que la fecha de entrega sea valida y futura
    try:
        fecha = parse_fecha(form["fecha_entrega"])
        if fecha < datetime.now():
            errores.append("La fecha de entrega debe ser futura")
    except Exception:
        errores.append("La fecha de entrega no es válida")

    return errores


# Validaciones extra para los comentarios
def validar_commentario_name(nombre):
    if not nombre:
        return False
    s = nombre.strip()
    return 3 <= len(s) <= 80

def validar_commentario_text(texto):
    if not texto:
        return False
    s = texto.strip()
    return len(s) >= 5

def validar_comentario_payload(payload):
    # Toma un diccionario con los datos del comentario y entrenga errores si los hay
    nombre = (payload.get("nombre") or "").strip()
    texto  = (payload.get("texto") or "").strip()
    errores = {}
    if not validar_commentario_name(nombre):
        errores["nombre"] = "Largo 3 a 80."
    if not validar_commentario_text(texto):
        errores["texto"] = "Mínimo 5 caracteres."
    return nombre, texto, errores